File: src/main/java/com/example/cqrs/command/entity/event/MoneyWithdrawnEventEntity.java
package com.example.cqrs.command.entity.event;

import com.example.cqrs.command.entity.event.metadata.EventMetadata;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("MoneyWithdrawnEvent")
@Entity
public class MoneyWithdrawnEventEntity extends AbstractAccountEventEntity {

    public MoneyWithdrawnEventEntity(String accountId, LocalDateTime eventDate,
                                     double amount, EventMetadata metadata) {
        super(accountId, eventDate, amount, metadata);
    }

    // factory method
    public static MoneyWithdrawnEventEntity of(String accountId, LocalDateTime now, double amount, EventMetadata eventMetadata) {
        return new MoneyWithdrawnEventEntity(accountId, now, amount, eventMetadata);
    }

}

File: src/main/java/com/example/cqrs/infrastructure/eventstore/entity/event/money/MoneyWithdrawnEventEntity.kt
package com.example.cqrs.infrastructure.eventstore.entity.event.money

import com.example.cqrs.infrastructure.eventstore.entity.base.AccountEventBaseEntity
import com.example.cqrs.infrastructure.eventstore.entity.base.metadata.EventMetadata
import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity
import java.time.LocalDateTime

/**
 * 출금 이벤트
 */
@DiscriminatorValue("MoneyWithdrawnEvent")
@Entity
class MoneyWithdrawnEventEntity(
    accountId: String,
    amount: Double,
    eventDate: LocalDateTime,
    metadata: EventMetadata
) : AccountEventBaseEntity(
    id = null,
    accountId = accountId,
    amount = amount,
    eventDate = eventDate,
    metadata = metadata
) {
    companion object {
        fun of(
            accountId: String,
            amount: Double,
            eventDate: LocalDateTime,
            metadata: EventMetadata
        ): MoneyWithdrawnEventEntity {
            return MoneyWithdrawnEventEntity(
                accountId = accountId,
                amount = amount,
                eventDate = eventDate,
                metadata = metadata
            )
        }
    }
}